src/twisted/internet/endpoints.py

```diff
@@ -2299,7 +2313,9 @@ def _parseClientTLS(
         ),
         clientFromString(reactor, endpoint)
         if endpoint is not None
-        else HostnameEndpoint(reactor, _idnaBytes(host), port, timeout, bindAddress),
+        else HostnameEndpoint(
+            reactor, _idnaBytes(host), port, timeout, (bindAddress, 0)
```

Comment:
This precludes binding a client port with a string, which we should probably address with a follow-up somehow.

Reply:
I don't understand. The `:`-joined format for TCP endpoints (which is embedded in the format for TLS client endpoints) doesn't have room for a port, only the binding address as a string/bytes. So previously we were effectively always binding the socket to (string, 0), and this makes that explicit.

(Looking at it again, I think my changes to `HostnameEndpoint.__init__` mean that this change to `_parseClientTLS` is unnecessary. If we didn't do either, then we hit the crash that I observed. I made the `_parseClientTLS` change first, which fixed my most immedaite problem, but really the `HostnameEndpoint` fix is the most complete one. I think I'll leave them both in place, though).

To "[bind] a client port with a string", you use something like `tls:..:endpoint=\"example.org:443:bindAddress=1.2.3.4\"`, and the `1.2.3.4` string is glued into a `("1.2.3.4", 0)` tuple and given to `HostnameEndpoint`.

What am I missing?

Reply:
> So previously we were effectively always binding the socket to (string, 0), and this makes that explicit

Yes, sorry, I didn't mean to say that this made things *worse*, just that it makes the restriction explicit, and having *noted* that restriction, we should lift it at some point.  We probably want to do it with a separate `bindPort` parameter or something, rather than escaping a colon in here.